be/src/util/quantile_state.cpp

```diff
@@ -0,0 +1,377 @@
+        return false;
+    }
+
+    const uint8_t* ptr = (uint8_t*)slice.data;
+    compression = *reinterpret_cast<const float*>(ptr);
+    ptr += sizeof(float);
+    // first byte : type
+    _type = (QuantileStateType)*ptr++;
+    switch (_type) {
+    case EMPTY:
+        // 1: empty 
+        break;
+    case SINGLE: {
+        // 2: single_data value
+        _single_data = *reinterpret_cast<const T*>(ptr);
+        ptr += sizeof(T);
+        break;
+    }
+    case EXPLICIT: {
+        // 3: number of explicit values
+        // make sure that num_explicit is positive
+        uint16_t num_explicits = decode_fixed16_le(ptr);
+        ptr += sizeof(uint16_t);
+        _explicit_data.reserve(std::min(num_explicits*2, QUANTILE_STATE_EXPLICIT_NUM));
+        _explicit_data.resize(num_explicits);
+        memcpy(&_explicit_data[0], ptr, num_explicits*sizeof(T));
+        ptr += num_explicits*sizeof(T);
+        break;
+    }
+    case TDIGEST: {
+        // 4: Tdigest object value
+        tdigest_ptr = new TDigest(0);
+        tdigest_ptr->unserialize(ptr);
+        break;
+    }
+    default:
+        // revert type to EMPTY
+        _type = EMPTY;
+        return false;
+    }
+    return true;
+
+}
+
+template<typename T>
+size_t QuantileState<T>::serialize(uint8_t* dst) const{
+    uint8_t* ptr = dst;
+    *reinterpret_cast<float *>(ptr) = compression;
+    ptr += sizeof(float);
+    switch (_type) {
+    case EMPTY: {
+        *ptr++ = EMPTY;
+        break;
+    }
+    case SINGLE: {
+        *ptr++ = SINGLE;
+        *reinterpret_cast<T *>(ptr) = _single_data;
+        ptr += sizeof(T);
+        break;
+    }
+    case EXPLICIT: {
+        *ptr++ = EXPLICIT;
+        uint16_t size = _explicit_data.size();
+        *reinterpret_cast<uint16_t *>(ptr) = size;
+        ptr += sizeof(uint16_t);
+        memcpy(ptr, &_explicit_data[0], size*sizeof(T));
+        ptr += size*sizeof(T);
+        break;
+    }
+    case TDIGEST: {
+        *ptr++ = TDIGEST;
+        size_t tdigest_size = tdigest_ptr->serialize(ptr);
+        ptr += tdigest_size;
+        break;
+    }
+    default:
+        break;
+    }
+    return ptr - dst;
+}
+
+template<typename T>
+void QuantileState<T>::merge(QuantileState<T>& other) {
+    switch(other._type) {
+    case EMPTY:
+        break;
+    case SINGLE: {
+        add_value(other._single_data);
+        break;
+    }
+    case EXPLICIT: {
+        switch(_type) {
+        case EMPTY:
+            _type = EXPLICIT;
+            _explicit_data.swap(other._explicit_data);
+            break;
+        case SINGLE:
+            _type = EXPLICIT;
+            _explicit_data.swap(other._explicit_data);
+            add_value(_single_data);
+            break;
+        case EXPLICIT:
+            if (_explicit_data.size() + other._explicit_data.size() > QUANTILE_STATE_EXPLICIT_NUM) {
+                _type = TDIGEST;
+                tdigest_ptr = new TDigest(compression);
```

Comment:
use `new TDigest(compression,compression,compression)` will save lots of memory.